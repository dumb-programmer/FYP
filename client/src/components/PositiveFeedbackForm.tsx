import { zodResolver } from "@hookform/resolvers/zod";
import { useForm } from "react-hook-form";
import { z } from "zod";

const options = ["correct", "easy-to-understand", "complete", "other"] as const;

const schema = z.object({ category: z.enum(options), comments: z.string().max(100) });
export default function PositiveFeedbackForm({ id }: { id: string }) {
    const { handleSubmit, watch, register } = useForm({
        defaultValues: {
            category: options[0],
            comments: ""
        },
        resolver: zodResolver(schema)
    });

    const onSubmit = handleSubmit(async (data) => {
        console.log(data);
    });

    return <form id={id} className="flex flex-col gap-6" onSubmit={onSubmit}>
        <div className="form-control flex flex-wrap gap-4 flex-row text-sm">
            {
                options.map(option => <label key={option} className={`flex items-center gap-2 relative bg-gray-200 p-4 rounded-xl w-fit cursor-pointer hover:bg-gray-300 transition-all ${watch("category") === option ? "bg-slate-800 text-white hover:bg-slate-700" : ""}`}>
                    <input type="radio" className="radio radio-primary invisible absolute" value={option} {...register("category", { required: true })} />
                    {option.split("-").join(" ")}
                </label>)
            }
        </div>
        <textarea className="textarea input-bordered" placeholder="(Optional) Something else you'd like to say"  {...register("comments")} ></textarea>
    </form>
}