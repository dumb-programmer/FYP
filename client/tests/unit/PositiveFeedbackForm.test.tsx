import React from "react";
import { render, screen } from "@testing-library/react";
import { expect, describe, it, vi } from "vitest";
import PositiveFeedbackForm from "../../src/components/PositiveFeedbackForm";
import userEvent from "@testing-library/user-event";

const id = "positive-feedback-form";
const messageId = "abc";
const onSuccessMock = vi.fn();
const renderComponent = () => render(<PositiveFeedbackForm id={id} messageId={messageId} onSuccess={onSuccessMock} />);

describe("PositiveFeebackForm", () => {
    it("renders correctly", () => {
        const { container } = renderComponent();

        expect(container).toMatchSnapshot();
    });

    it("comments input works", async () => {
        renderComponent();

        const input = screen.getByRole("textbox");

        await userEvent.type(input, "hello");

        expect(input.value).toEqual("hello");
    });

    it("comments input works", async () => {
        renderComponent();

        const input = screen.getByRole("textbox");

        await userEvent.type(input, "hello");

        expect(input.value).toEqual("hello");
    });

    it("radio inputs work", async () => {
        renderComponent();

        const radioInputs = screen.getAllByRole("radio");

        for (const radioInput of radioInputs) {
            await userEvent.click(radioInput);

            expect(radioInput).toBeChecked();
        }
    });
})