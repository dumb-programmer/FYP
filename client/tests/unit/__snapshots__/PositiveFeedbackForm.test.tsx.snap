// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PositiveFeebackForm > renders correctly 1`] = `
<div>
  <form
    class="flex flex-col gap-6"
    id="positive-feedback-form"
  >
    <div
      class="form-control flex flex-wrap gap-4 flex-row text-sm"
    >
      <label
        class="flex items-center gap-2 relative bg-gray-200 p-4 rounded-xl w-fit cursor-pointer hover:bg-gray-300 transition-all bg-slate-800 text-white hover:bg-slate-700"
      >
        <input
          class="radio radio-primary invisible absolute"
          name="category"
          type="radio"
          value="correct"
        />
        correct
      </label>
      <label
        class="flex items-center gap-2 relative bg-gray-200 p-4 rounded-xl w-fit cursor-pointer hover:bg-gray-300 transition-all "
      >
        <input
          class="radio radio-primary invisible absolute"
          name="category"
          type="radio"
          value="easy-to-understand"
        />
        easy to understand
      </label>
      <label
        class="flex items-center gap-2 relative bg-gray-200 p-4 rounded-xl w-fit cursor-pointer hover:bg-gray-300 transition-all "
      >
        <input
          class="radio radio-primary invisible absolute"
          name="category"
          type="radio"
          value="complete"
        />
        complete
      </label>
      <label
        class="flex items-center gap-2 relative bg-gray-200 p-4 rounded-xl w-fit cursor-pointer hover:bg-gray-300 transition-all "
      >
        <input
          class="radio radio-primary invisible absolute"
          name="category"
          type="radio"
          value="other"
        />
        other
      </label>
    </div>
    <textarea
      class="textarea input-bordered"
      name="comments"
      placeholder="(Optional) Something else you'd like to say"
    />
  </form>
</div>
`;
